### PR TITLE
Link ancestral sequences correctly when gathering Align Slices

### PR DESCRIPTION
**Pull request by JAlvarezJarreta**
## Description

EPO alignments have been displayed incorrectly for ages (most likely since e86). For instance, there were cases when one or more species were displayed twice [[link](http://nov2020.archive.ensembl.org/Pongo_abelii/Location/Compara_Alignments/Image?align=1628&db=core&r=9%3A133995000-134052000)] or the ancestral sequences were displaying the wrong information [[link](https://www.ensembl.org/Pongo_abelii/Location/Compara_Alignments/Image?align=1944&db=core&r=9:133995000-134053000)] (see the left part of the ancestral sequence above gorilla). This PR fixes this issue.

**Related JIRA tickets:**
- ENSCOMPARASW-4081

## Overview of changes
- Fixed the bugs located in `AlignSliceAdaptor.pm`:
    - The ancestral sequences were merged without checking whether they were parent of the same species or not. Now that is resolved not merging them unless they are linked by `right_node_id == node_id`.
    - If two trees had differed in the species order, the method would duplicate these species. I have decided to preserve the order of the first tree when conflicts are found, and just merge the information of the species (and their ancestral sequences) properly.
- Updated PODs for both `scripts/dumps/DumpAlignSlice.pl` and `modules/Bio/EnsEMBL/Compara/DBSQL/AlignSliceAdaptor.pm`, and code clean-up in the latter.
- The `DEBUG info` code has been quite useful, so I have updated it in case it is required again in the future (made it easier to print more information if required).
- Changed default `dbname` in `DumpAlignSlice.pl` to `compara_curr` as it is what we will use most times.

## Testing
I have tested the script in both `release/104` and this branch for the issues found in primates EPO and murinae EPO. For the former:
```
perl $ENSEMBL_CVS_ROOT_DIR/ensembl-compara/scripts/dumps/DumpAlignSlice.pl --species pongo_abelii --seq_region 9 --seq_region_start 133995000 --seq_region_end 134053000 --alignment_type EPO --species_set_name primates --reg_conf $COMPARA_REG_PATH
```
If you compare both output files, you will notice the `ancestral_sequence` for `gorilla_gorilla` was including an alignment when there is none for the actual species, and now that section is located in the `ancestral_sequence` for `homo_sapiens` (its correct place).
And the latter:
```
perl $ENSEMBL_CVS_ROOT_DIR/ensembl-compara/scripts/dumps/DumpAlignSlice.pl --species mus_musculus --seq_region 4 --seq_region_start 136370000 --seq_region_end 136450000 --alignment_type EPO --species_set_name murinae --reg_conf $COMPARA_REG_PATH
```
In this case the species only appear once, instead of multiple times, and their information is combined as expected.

## Notes
Although I'm well known to be quite picky with the indentation, I have tried to leave as much of the previous code as untouched as possible so it is easier to track the actual important changes.

### CODE REVIEW

**Comment by JAlvarezJarreta** [comment]
Production has requested this to be reported as a known bug for e103 and fixed in e104 instead.

**Review by carlacummins — APPROVED** [review]

File: modules/Bio/EnsEMBL/Compara/DBSQL/AlignSliceAdaptor.pm [hunk 1]
```diff
@@ -433,111 +381,101 @@ sub _combine_genomic_align_trees {
...
+      $species_genome_db = $species->{genome_db} if defined $species;
+      $species_right_node_id = $species->{right_node_id} if defined $species;
       $match = 1;
 
-      ## 1. Use info from species_right_node_id if available
+      ## 1. Use info from $species_right_node_id if available
       if (defined($species_right_node_id) and $species_right_node_id == $this_node_id) {
-          $species_order->[$species_counter]->{right_node_id} = $this_right_node_id;
-          # #         $species_order->[$species_counter]->{last_node} = $this_genomic_align_node;
+        $species_order->[$species_counter]->{right_node_id} = $this_right_node_id;
```
Comment (CristiGuijarro) [inline modules/Bio/EnsEMBL/Compara/DBSQL/AlignSliceAdaptor.pm:395]: There are a few of these git blame/white space change issues floating about. I don't mind personally, but I think we are supposed to preserve where possible.

File: modules/Bio/EnsEMBL/Compara/DBSQL/AlignSliceAdaptor.pm [hunk 2]
```diff
@@ -433,111 +381,101 @@ sub _combine_genomic_align_trees {
...
         push (@{$species_order->[$species_counter]->{genomic_align_ids}}, @$these_genomic_align_ids);
+        if (defined $pending_ancestral_species) {
+          my $prev_species_index = $species_counter - 1;
+          $species_order->[$prev_species_index]->{right_node_id} = $pending_ancestral_species->{right_node_id};
+          push @{$species_order->[$prev_species_index]->{genomic_align_ids}}, @{$pending_ancestral_species->{genomic_align_ids}};
+          undef $pending_ancestral_species;
+        }
         ## DEBUG info
```
Comment (CristiGuijarro) [inline modules/Bio/EnsEMBL/Compara/DBSQL/AlignSliceAdaptor.pm:403]:
```suggestion
```
May as well delete this too, since `_debug_info` as sub name is pretty self explanatory.

File: modules/Bio/EnsEMBL/Compara/DBSQL/AlignSliceAdaptor.pm [hunk 3]
```diff
@@ -433,111 +381,101 @@ sub _combine_genomic_align_trees {
...
+          $species_order->[$species_counter]->{right_node_id} = $this_right_node_id;
+          push (@{$species_order->[$species_counter]->{genomic_align_ids}}, @$these_genomic_align_ids);
+          if (defined $pending_ancestral_species) {
+            my $prev_species_index = $species_counter - 1;
+            $species_order->[$prev_species_index]->{right_node_id} = $pending_ancestral_species->{right_node_id};
+            push @{$species_order->[$prev_species_index]->{genomic_align_ids}}, @{$pending_ancestral_species->{genomic_align_ids}};
+            undef $pending_ancestral_species;
+          }
+        }
+        ## DEBUG info
```
Comment (CristiGuijarro) [inline modules/Bio/EnsEMBL/Compara/DBSQL/AlignSliceAdaptor.pm:427]:
```suggestion
```
Unnecessary comment

File: modules/Bio/EnsEMBL/Compara/DBSQL/AlignSliceAdaptor.pm [hunk 4]
```diff
@@ -433,111 +381,101 @@ sub _combine_genomic_align_trees {
...
+        $species_counter = $existing_species_names{$this_genome_db->name};
         $species_order->[$species_counter]->{right_node_id} = $this_right_node_id;
         push (@{$species_order->[$species_counter]->{genomic_align_ids}}, @$these_genomic_align_ids);
+        if (defined $pending_ancestral_species) {
+          my $prev_species_index = $species_counter - 1;
+          $species_order->[$prev_species_index]->{right_node_id} = $pending_ancestral_species->{right_node_id};
+          push @{$species_order->[$prev_species_index]->{genomic_align_ids}}, @{$pending_ancestral_species->{genomic_align_ids}};
+          undef $pending_ancestral_species;
+        }
         ## DEBUG info
```
Comment (CristiGuijarro) [inline modules/Bio/EnsEMBL/Compara/DBSQL/AlignSliceAdaptor.pm:442]:
```suggestion
```
As the rest

File: modules/Bio/EnsEMBL/Compara/DBSQL/AlignSliceAdaptor.pm [hunk 5]
```diff
@@ -433,111 +381,101 @@ sub _combine_genomic_align_trees {
...
+          }
+          splice(@$species_order, $species_counter, 0, {
+            genome_db         => $this_genome_db,
+            right_node_id     => $this_right_node_id,
+            genomic_align_ids => [ @$these_genomic_align_ids ],
           });
+          # Update the existing species names
+          $existing_species_names{$this_genome_db->name} = $species_counter;
+        }
         ## DEBUG info
```
Comment (CristiGuijarro) [inline modules/Bio/EnsEMBL/Compara/DBSQL/AlignSliceAdaptor.pm:469]:
```suggestion
```

File: scripts/dumps/DumpAlignSlice.pl [hunk 6]
```diff
@@ -85,7 +76,7 @@ =head2 GENERAL CONFIGURATION
 =item B<[--dbname compara_db_name]>
 
 the name of compara DB in the registry_configuration_file or any
-of its aliases. Uses "compara" by default.
+of its aliases. Default is "compara_curr".
```
Comment (CristiGuijarro) [inline scripts/dumps/DumpAlignSlice.pl:79]:
```suggestion
of its aliases.
Default "compara_curr"
```

File: scripts/dumps/DumpAlignSlice.pl [hunk 7]
```diff
@@ -103,15 +94,19 @@ =head2 SPECIFYING THE QUERY SLICE
 
 =item B<--seq_region region_name>
 
-Query region name, i.e. the chromosome name
+Query region name, i.e. the chromosome name. Default 13.
```
Comment (CristiGuijarro) [inline scripts/dumps/DumpAlignSlice.pl:97]:
```suggestion
Query region name, i.e. the chromosome name.
Default 13
```
Just for readability. Full stop removed for consistency.

File: scripts/dumps/DumpAlignSlice.pl [hunk 8]
```diff
@@ -103,15 +94,19 @@ =head2 SPECIFYING THE QUERY SLICE
...
+Default 32906420
+
 =item B<--seq_region_end end>
 
+Default 32906519
+
 =item B<[--seq_region_strand strand]>
 
-The strand of the query. It can be either +1 or -1. Default is -1.
+The strand of the query. It can be either +1 or -1. Default is +1.
```
Comment (CristiGuijarro) [inline scripts/dumps/DumpAlignSlice.pl:109]:
```suggestion
The strand of the query. It can be either +1 or -1.
Default +1
```

File: scripts/dumps/DumpAlignSlice.pl [hunk 9]
```diff
@@ -121,7 +116,7 @@ =head2 SPECIFYING THE ALIGNMENT TYPE
 
 =item B<[--alignment_type method_link_name]>
 
-The type of alignment. Default is "BLASTZ_NET"
+The type of alignment. Default is "EPO".
```
Comment (CristiGuijarro) [inline scripts/dumps/DumpAlignSlice.pl:119]:
```suggestion
The type of alignment.
Default "EPO"
```

File: scripts/dumps/DumpAlignSlice.pl [hunk 10]
```diff
@@ -130,12 +125,18 @@ =head2 SPECIFYING THE ALIGNMENT TYPE
 should correspond to the name of the core database in the
 registry_configuration_file or any of its aliases
 
+=item B<[--species_set_name name]>
+
+Pre-defined name for the set of species used. For multiple alignment sets only.
+eg "mammals" or "primates" for EPO alignments; "amniotes" for PECAN alignments.
+Default is "mammals".
```
Comment (CristiGuijarro) [inline scripts/dumps/DumpAlignSlice.pl:132]:
```suggestion
Default "mammals"
```

File: scripts/dumps/DumpAlignSlice.pl [hunk 11]
```diff
@@ -194,17 +195,19 @@ =head2 OUTPUT
         ~/.ensembl_init will be used.
     [--dbname compara_db_name]
         the name of compara DB in the registry_configuration_file or any
-        of its aliases. Uses "compara" by default.
+        of its aliases. Default is "compara_curr".
```
Comment (CristiGuijarro) [inline scripts/dumps/DumpAlignSlice.pl:198]:
```suggestion
        of its aliases.
        Default "compara_curr"
```

File: scripts/dumps/DumpAlignSlice.pl [hunk 12]
```diff
@@ -194,17 +195,19 @@ =head2 OUTPUT
...
 
   For the query slice:
     [--species query_species]
         Query species. Default is "human"
     [--coord_system coordinates_name]
         Query coordinate system. Default is "chromosome"
     --seq_region region_name
-        Query region name, i.e. the chromosome name
+        Query region name, i.e. the chromosome name. Default 13.
```
Comment (CristiGuijarro) [inline scripts/dumps/DumpAlignSlice.pl:206]:
```suggestion
        Query region name, i.e. the chromosome name.
        Default 13
```

File: modules/Bio/EnsEMBL/Compara/DBSQL/AlignSliceAdaptor.pm [hunk 13]
```diff
@@ -433,111 +381,101 @@ sub _combine_genomic_align_trees {
...
+          };
+          $debug_msg = "TENTATIVE " . $debug_msg;
+        } else {
+          $species_order->[$species_counter]->{right_node_id} = $this_right_node_id;
+          push (@{$species_order->[$species_counter]->{genomic_align_ids}}, @$these_genomic_align_ids);
+          if (defined $pending_ancestral_species) {
+            my $prev_species_index = $species_counter - 1;
+            $species_order->[$prev_species_index]->{right_node_id} = $pending_ancestral_species->{right_node_id};
+            push @{$species_order->[$prev_species_index]->{genomic_align_ids}}, @{$pending_ancestral_species->{genomic_align_ids}};
+            undef $pending_ancestral_species;
```
Comment (CristiGuijarro) [inline modules/Bio/EnsEMBL/Compara/DBSQL/AlignSliceAdaptor.pm:424]: It's hard to follow because of varying indentations (I know, not your fault) - just want to check and take your word for it that `$pending_ancestral_sequences` is not trapped in an always `undef` state! Also to double check that `undef $pending_ancestral_sequences` does as you are intending it to do. `undef` like this seems to work in unexpected ways!

**Review by CristiGuijarro — APPROVED** [review]
I am approving regardless of the various formatting comments I have made, because none of those are necessary to the functionality of the code. And approving based on your thorough testing and word that it works based on my single function-related comment :+1:

File: modules/Bio/EnsEMBL/Compara/DBSQL/AlignSliceAdaptor.pm (hunk 13 above)
Comment (JAlvarezJarreta) [reply modules/Bio/EnsEMBL/Compara/DBSQL/AlignSliceAdaptor.pm:424]: In all my testing it worked as expected: when the species is an ancestral sequence, unless the `right_node_id` is found on the tree, it is stored in `$pending_ancestral_sequences` to see if the next species in the tree matched the next one in `$species_order` (to aggregate ancestral sequences accordingly). I also feared `undef` way delete the content (so I was cautious and tested it thoroughly) but since it is a reference, it seems it is dereferencing it only (and since the data has been copied somewhere else, it is not deleted).

File: modules/Bio/EnsEMBL/Compara/DBSQL/AlignSliceAdaptor.pm (hunk 1 above)
Comment (JAlvarezJarreta) [reply modules/Bio/EnsEMBL/Compara/DBSQL/AlignSliceAdaptor.pm:395]: I know, and I tried, but they really did my work harder to track and follow (2 spaces as indentation is already hard enough :confounded:), so I'm willing to take the blame :nerd_face:
